Implement instanceID method

irozzo-1A opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR implements the cloud provider `InstanceID` method that is called by the [node controller](https://github.com/kubernetes/kubernetes/blob/3a51aaf8b4baff1242ea30b08d599d913fe7a127/staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go#L522). This allows the users of this CCM not to specify the `provider-id` kubelet flag.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #74 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

irozzo-1A commented:
/assign @RainbowMango

irozzo-1A commented on pkg/cloudprovider/huaweicloud/instances_test.go:
```diff
@@ -68,3 +77,90 @@ func TestParseInstanceTypeFromServerInfo(t *testing.T) {
+	if err != nil {
+		t.Fatalf("Error occurred while marshaling the response: %v", err)
+	}
+	// Handle server creation requests.
```
change to `Handle server list requests.`

Dingshujie commented on pkg/cloudprovider/huaweicloud/instances.go:
```diff
@@ -264,3 +228,93 @@ func (i *Instances) parseInstanceTypeFromServerInfo(srv *servers.Server) (string
+		return nil, ErrNotFound
+	}
+
+	return &serverList[0], nil
```
servers can have same hostname, can't guarantee serverList[0]  is correct one

RainbowMango commented:
@irozzo-1A 
Thanks for your PR. Have you tested it locally?

irozzo-1A commented:
> @irozzo-1A
> Thanks for your PR. Have you tested it locally?

@RainbowMango I tested it on a cluster I created, but not locally following [this](https://github.com/kubernetes-sigs/cloud-provider-huaweicloud/blob/master/docs/quick-start-with-local-cluster.md) doc.  Node initialisation went through smoothly. I'm still struggling with an authorization issue when trying to create a service of type LoadBalancer, I don't know yet if it's a configuration issue or a bug, but in case I will file another issue. I don't see any evident relationship with my change though. 

Regarding this PR there is something that I would like to check with you. [Here](https://kubernetes.io/docs/concepts/cluster-administration/cloud-providers/#node-name-4) is mentioned that `--hostname-override` should be set to the node private IP. What is the reason behind? I think that this does not play nice with this feature, because the node name should correspond to the actual server name for it to work.

irozzo-1A commented:
@RainbowMango Small update. I tested it using `hack/local-up-cluster.sh` following the quickstart doc and I confirm that the node get initialized successfully, so no regressions I would say. The only caveat as I anticipated is that if the `--provider-id` flag is not specified the node name should match the server name, so `--hostname-override` should not be specified or being equal to the server name.

RainbowMango commented:
I just quickly tested it using `hack/local-up-cluster.sh` and set the environment as per [the doc](https://github.com/kubernetes-sigs/cloud-provider-huaweicloud/blob/master/docs/quick-start-with-local-cluster.md#set-environment), except that I removed two variables:
- #export HOSTNAME_OVERRIDE=xxx
- #export KUBELET_PROVIDER_ID=xxx

I just found the node will be removed by node lifecycle controller:
```
E0702 11:04:10.183963    5918 node_controller.go:428] failed to get instance ID from cloud provider: instance not found
E0702 11:04:11.399815    5918 node_controller.go:428] failed to get instance ID from cloud provider: instance not found
I0702 11:04:11.930116    5918 instances.go:175] InstanceShutdownByProviderID is called. input provider ID: 
E0702 11:04:12.095204    5918 node_controller.go:428] failed to get instance ID from cloud provider: instance not found
E0702 11:04:12.656009    5918 instances.go:179] Get server info failed. provider id: , error: error occurred while getting server with server id: , error: {"ErrorCode":"APIGW.0101","Message":"The API does not exist or has not been published in the environ}
E0702 11:04:12.656035    5918 node_lifecycle_controller.go:154] error checking if node 192.168.1.122 is shutdown: error occurred while getting server with server id: , error: {"ErrorCode":"APIGW.0101","Message":"The API does not exist or has not been publ}
I0702 11:04:13.077677    5918 node_lifecycle_controller.go:180] deleting node since it is no longer present in cloud provider: 192.168.1.122
```

Kubernetes version:
```
Server Version: version.Info{Major:"1", Minor:"19+", GitVersion:"v1.19.0-beta.2.577+180af4240c4290-dirty", GitCommit:"180af4240c42906f072112e7ac7818f4b3574c5a", GitTreeState:"dirty", BuildDate:"2020-07-02T03:03:15Z", GoVersion:"go1.14.4", Compiler:"gc", P}
```

/cc @Dingshujie 
Involve @Dingshujie here, she is an expert with a lot of production experience.

RainbowMango commented on pkg/cloudprovider/huaweicloud/instances.go:
```diff
@@ -100,9 +110,17 @@ func (i *Instances) NodeAddressesByProviderID(ctx context.Context, providerID st
+	srv, err := i.getServerByName(nodeName)
+	if err != nil {
+		if err == ErrNotFound {
+			return "", cloudprovider.InstanceNotFound
```
I'd prefer adding one error log when returns an error.

Same as below.

irozzo-1A replied:
note that it gets eventually logged by the node controller, e.g.

```
E0702 07:24:11.537080   20857 node_controller.go:428] failed to get instance ID from cloud provider: instance not found
```

Personally I would find it a bit redundant in this case, but it does not cost me much effort to add it if you still prefer ;-)

RainbowMango replied:
It's no big deal. No worries.

irozzo-1A commented:
I just quickly tested it using hack/local-up-cluster.sh and set the environment as per the doc, except that I removed two variables:

#export HOSTNAME_OVERRIDE=xxx
#export KUBELET_PROVIDER_ID=xxx

@RainbowMango sorry if I did not mention it, but this is not enough to test it with `hack/local-up-cluster.sh` because if you don't specify those env variables `--hostname-override` [defaults to `127.0.0.1`](https://github.com/kubernetes/kubernetes/blob/908847c/hack/local-up-cluster.sh#L77) and `--provider-id` [defaults to the hostname](https://github.com/kubernetes/kubernetes/blob/908847c/hack/local-up-cluster.sh#L83). If you want to test it that way you should do the following:

- Removed those lines: https://github.com/kubernetes/kubernetes/blob/908847c/hack/local-up-cluster.sh#L712,L713
- `export HOSTNAME_OVERRIDE="$(hostname)"`

RainbowMango commented:
Yeah, you are right.
/lgtm
/approve

Thanks again for your contribution :). Just feel free to send a PR if you want to improve it.
